File: solutions/ruby/perfect-numbers/1/perfect_numbers.rb
class PerfectNumber
  def self.classify n
    raise RuntimeError unless n.positive?

    aliquot_sum = aliquot_sum n
    return 'deficient' if aliquot_sum < n
    return 'abundant' if aliquot_sum > n
    return 'perfect'
  end

  private_class_method def self.aliquot_sum n
    (1...n).filter { |x| n % x == 0 }.sum
  end
end

File: solutions/lua/perfect-numbers/1/perfect-numbers.lua
local function aliquot_sum(n)
    local sum = 0
    for f = 1, (n >> 1) do
        if n % f == 0 then sum = sum + f end
    end
    return sum
end

local function classify(n)
    local s = aliquot_sum(n)
    return s == n and 'perfect' or s > n and 'abundant' or 'deficient'
end

return {
  aliquot_sum = aliquot_sum,
  classify = classify
}